File: app/src/main/java/com/qurantv/app/data/repo/TimingRepository.kt
package com.qurantv.app.data.repo

import com.qurantv.app.data.api.AyahTimingDto
import com.qurantv.app.data.api.Mp3QuranApi
import com.qurantv.app.data.api.SoarDto
import com.qurantv.app.data.api.TimingReadDto
import com.qurantv.app.data.api.toDomain
import com.qurantv.app.data.cache.JsonDiskCache
import com.qurantv.app.domain.CatalogParsing
import com.qurantv.app.domain.SurahTiming
import com.qurantv.app.domain.TimingAccuracy
import com.qurantv.app.domain.TimingRead
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json

/**
 * Per-ayah timing data (PROMPT.md Parts 3.6–3.8):
 *  - the `reads` list is fetched once at startup and cached forever;
 *  - per-(read, surah) timing is immutable → cached forever;
 *  - reads are matched to reciter moshafs by `folder_url` ↔ `server`
 *    (normalized trailing slashes) — timing read ids are NOT reciter ids.
 */
class TimingRepository(
    private val api: Mp3QuranApi,
    private val cache: JsonDiskCache,
    private val json: Json,
) {

    suspend fun reads(): List<TimingRead> {
        val key = "reads"
        return cache.singleFlight(key) {
            val cached = cache.read(JsonDiskCache.TIMING, key)
            if (cached != null) {
                return@singleFlight json.decodeFromString<List<TimingReadDto>>(cached).map { it.toDomain() }
            }
            val list = api.timingReads().map { it.toDomain() }
            runCatching {
                cache.write(JsonDiskCache.TIMING, key, json.encodeToString(list))
            }
            list
        }
    }

    /** The read whose folder matches this moshaf server, or null when untimed. */
    suspend fun readForMoshaf(server: String): TimingRead? {
        val target = CatalogParsing.normalizeServerUrl(server)
        return reads().firstOrNull { CatalogParsing.normalizeServerUrl(it.folderUrl) == target }
    }

    /** Normalized folder URLs of every read that has ayah timing. */
    suspend fun timedServerUrls(): Set<String> =
        reads().map { CatalogParsing.normalizeServerUrl(it.folderUrl) }.toSet()

    /**
     * The surah ids that have per-ayah timing files for this read (the
     * `ayat_timing/soar` list — some reads cover fewer than all 114 surahs).
     * Immutable → cached forever. Returns null on failure so callers can fall
     * back to showing the full list (graceful degradation).
     */
    suspend fun surahsWithTiming(readId: Int): Set<Int>? {
        val key = "soar_r$readId"
        return cache.singleFlight(key) {
            val cached = cache.read(JsonDiskCache.TIMING, key)
            if (cached != null) {
                return@singleFlight json.decodeFromString<List<SoarDto>>(cached).map { it.id }.toSet()
            }
            try {
                val list = api.soar(readId)
                runCatching { cache.write(JsonDiskCache.TIMING, key, json.encodeToString(list)) }
                list.map { it.id }.toSet()
            } catch (e: Exception) {
                null // unknown — callers keep the full surah list
            }
        }
    }

    /**
     * Whether the (read, surah) timing is ACTUALLY usable — the soar list can
     * over-claim (e.g. read 122 البنا المجود lists surah 97, but the timing's
     * total is ~6% shorter than the real mp3, so sync is disabled at playback).
     * Verdict: the timing exists AND (when the mp3 is small enough to probe) the
     * mp3 duration matches the timing total within the accuracy tolerance.
     * Cached forever once computed. Returns null when the verdict is unknown
     * (mp3 too large to probe, or the probe failed).
     */
    suspend fun timingUsability(readId: Int, surahId: Int, mp3Url: String): Boolean? {
        val timing = timingFor(readId, surahId) ?: return false
        val verdictKey = "usable2_r${readId}_s$surahId"
        cache.read(JsonDiskCache.TIMING, verdictKey)?.let { return it == "1" }
        val usable = withContext(Dispatchers.IO) {
            probeMp3DurationMs(mp3Url)?.let { mp3Ms ->
                TimingAccuracy.isReliable(mp3Ms, timing.lastEndMs)
            }
        }
        if (usable != null) {
            runCatching { cache.write(JsonDiskCache.TIMING, verdictKey, if (usable) "1" else "0") }
        }
        return usable
    }

    /**
     * Probes an mp3's duration via MediaMetadataRetriever. Only files up to
     * [MAX_PROBE_BYTES] are probed (VBR mp3s without a Xing header require a
     * full scan, so huge surah files would cost too much to check eagerly).
     */
    private fun probeMp3DurationMs(url: String): Long? {
        return try {
            val size = runCatching {
                val conn = java.net.URL(url).openConnection() as java.net.HttpURLConnection
                conn.requestMethod = "HEAD"
                conn.setRequestProperty("Range", "bytes=0-0")
                conn.setRequestProperty("User-Agent", "QuranTv/1.0")
                val cl = conn.getHeaderField("Content-Range")?.substringAfter('/')?.toLongOrNull()
                conn.disconnect()
                cl
            }.getOrNull()
            if (size != null && size > MAX_PROBE_BYTES) return null
            val mmr = android.media.MediaMetadataRetriever()
            try {
                mmr.setDataSource(url)
                mmr.extractMetadata(android.media.MediaMetadataRetriever.METADATA_KEY_DURATION)?.toLongOrNull()
            } finally {
                runCatching { mmr.release() }
            }
        } catch (e: Exception) {
            null
        }
    }

    private companion object {
        /** Largest mp3 (bytes) we'll eagerly probe for timing accuracy. */
        const val MAX_PROBE_BYTES = 10L * 1024 * 1024
    }

    /** Timing for (read, surah); null on any failure → graceful degradation. */
    suspend fun timingFor(readId: Int, surahId: Int): SurahTiming? {
        val key = "s${surahId}_r$readId"
        return cache.singleFlight(key) {
            val cached = cache.read(JsonDiskCache.TIMING, key)
            if (cached != null) {
                return@singleFlight json.decodeFromString<List<AyahTimingDto>>(cached).toDomain(readId, surahId)
            }
            try {
                val list = api.ayahTiming(surahId, readId)
                runCatching {
                    cache.write(JsonDiskCache.TIMING, key, json.encodeToString(list))
                }
                list.toDomain(readId, surahId)
            } catch (e: Exception) {
                null // no timing for this pair — play without ayah sync
            }
        }
    }

    /** Warm the cache for the next surah while the current one plays. */
    fun prefetch(readId: Int, surahId: Int, scope: CoroutineScope) {
        scope.launch {
            timingFor(readId, surahId)
        }
    }
}

File: web/src/data/repo/TimingRepository.ts
import { normalizeServerUrl } from "../../domain/CatalogParsing";
import { isReliable } from "../../domain/TimingAccuracy";
import type { SurahTiming, TimingRead } from "../../domain/Models";
import {
  Mp3QuranApi,
  ayahTimingToDomain,
  timingReadDtoToDomain,
  type TimingIndex,
} from "../api/Mp3QuranApi";
import type { AyahTimingDto, SoarDto, TimingReadDto } from "../api/Dtos";
import { CACHE, JsonDiskCache } from "../cache/JsonDiskCache";

const MAX_PROBE_BYTES = 10 * 1024 * 1024;

export class TimingRepository {
  private timingIndexCache: TimingIndex | null = null;

  constructor(
    private readonly api: Mp3QuranApi,
    private readonly cache: JsonDiskCache,
  ) {}

  /** Load fast O(1) timing index from CDN/mirror if available. */
  async getTimingIndex(): Promise<TimingIndex | null> {
    if (this.timingIndexCache !== null) return this.timingIndexCache;
    const key = "timing_index_fast";
    return this.cache.singleFlight(key, async () => {
      const cached = await this.cache.read(CACHE.TIMING, key);
      if (cached !== null) {
        this.timingIndexCache = JSON.parse(cached) as TimingIndex;
        return this.timingIndexCache;
      }
      const idx = await this.api.timingIndex();
      if (idx) {
        await this.cache.write(CACHE.TIMING, key, JSON.stringify(idx));
        this.timingIndexCache = idx;
      }
      return idx;
    });
  }

  async reads(): Promise<TimingRead[]> {
    const key = "reads";
    return this.cache.singleFlight(key, async () => {
      const cached = await this.cache.read(CACHE.TIMING, key);
      if (cached !== null) {
        return (JSON.parse(cached) as TimingReadDto[]).map(timingReadDtoToDomain);
      }
      const dtos = await this.api.timingReads();
      await this.cache.write(CACHE.TIMING, key, JSON.stringify(dtos));
      return dtos.map(timingReadDtoToDomain);
    });
  }

  /** The read whose folder matches this moshaf server, or null when untimed. */
  async readForMoshaf(server: string): Promise<TimingRead | null> {
    const target = normalizeServerUrl(server);
    const all = await this.reads();
    const match = all.find((r) => normalizeServerUrl(r.folderUrl) === target);
    if (match) return match;

    // Fast path fallback: O(1) check index
    const idx = await this.getTimingIndex();
    if (idx && idx.servers[target]) {
      const record = idx.servers[target];
      return {
        id: record.read_id,
        name: "",
        rewaya: null,
        folderUrl: target,
        slug: null,
      };
    }
    return null;
  }

  /** Normalized folder URLs of every read that has ayah timing. */
  async timedServerUrls(): Promise<Set<string>> {
    const idx = await this.getTimingIndex();
    if (idx) {
      return new Set(Object.keys(idx.servers).map(normalizeServerUrl));
    }
    const all = await this.reads();
    return new Set(all.map((r) => normalizeServerUrl(r.folderUrl)));
  }

  /**
   * The surah ids that have per-ayah timing files for this read (the
   * ayat_timing/soar list). Cached forever.
   */
  async surahsWithTiming(readId: number, serverUrl?: string): Promise<Set<number> | null> {
    // Fast path via index
    if (serverUrl) {
      const idx = await this.getTimingIndex();
      const target = normalizeServerUrl(serverUrl);
      if (idx && idx.servers[target] && idx.servers[target].surahs) {
        const s = idx.servers[target].surahs;
        if (s === "all" || s === "*") {
          return new Set(Array.from({ length: 114 }, (_, i) => i + 1));
        }
        if (Array.isArray(s)) {
          return new Set(s);
        }
      }
    }

    const key = `soar_r${readId}`;
    return this.cache.singleFlight(key, async () => {
      const cached = await this.cache.read(CACHE.TIMING, key);
      if (cached !== null) {
        return new Set((JSON.parse(cached) as SoarDto[]).map((s) => s.id));
      }
      try {
        const list = await this.api.soar(readId);
        await this.cache.write(CACHE.TIMING, key, JSON.stringify(list));
        return new Set(list.map((s) => s.id));
      } catch {
        return null;
      }
    });
  }

  /**
   * Whether the (read, surah) timing is ACTUALLY usable.
   */
  async timingUsability(readId: number, surahId: number, mp3Url: string): Promise<boolean | null> {
    const timing = await this.timingFor(readId, surahId);
    if (timing === null) return false;
    const verdictKey = `usable2_r${readId}_s${surahId}`;
    const cachedVerdict = await this.cache.read(CACHE.TIMING, verdictKey);
    if (cachedVerdict !== null) return cachedVerdict === "1";
    const usable = await this.probeMp3DurationMs(mp3Url).then((mp3Ms) =>
      mp3Ms !== null ? isReliable(mp3Ms, timing.lastEndMs) : null,
    );
    if (usable !== null) {
      await this.cache.write(CACHE.TIMING, verdictKey, usable ? "1" : "0");
    }
    return usable;
  }

  private async probeMp3DurationMs(url: string): Promise<number | null> {
    try {
      const size = await this.headContentLength(url);
      if (size !== null && size > MAX_PROBE_BYTES) return null;
      return await new Promise<number | null>((resolve) => {
        const audio = new Audio();
        audio.preload = "metadata";
        const timer = setTimeout(() => {
          audio.src = "";
          resolve(null);
        }, 15_000);
        audio.onloadedmetadata = () => {
          clearTimeout(timer);
          const ms = Math.round(audio.duration * 1000);
          audio.src = "";
          resolve(Number.isFinite(ms) ? ms : null);
        };
        audio.onerror = () => {
          clearTimeout(timer);
          resolve(null);
        };
        audio.src = url;
      });
    } catch {
      return null;
    }
  }

  private async headContentLength(url: string): Promise<number | null> {
    try {
      const res = await fetch(url, { method: "HEAD" });
      const cr = res.headers.get("Content-Range");
      if (cr) {
        const total = Number.parseInt(cr.substring(cr.lastIndexOf("/") + 1), 10);
        if (Number.isFinite(total)) return total;
      }
      const cl = res.headers.get("Content-Length");
      return cl ? Number.parseInt(cl, 10) : null;
    } catch {
      return null;
    }
  }

  /** Timing for (read, surah); null on any failure → graceful degradation. */
  async timingFor(
    readId: number,
    surahId: number,
    slug?: string,
    reciterId?: number,
    moshafId?: number,
  ): Promise<SurahTiming | null> {
    const key = `s${surahId}_r${readId}${slug ? `_${slug}` : ""}`;
    return this.cache.singleFlight(key, async () => {
      const cached = await this.cache.read(CACHE.TIMING, key);
      if (cached !== null) {
        return ayahTimingToDomain(JSON.parse(cached) as (AyahTimingDto | number)[], readId, surahId);
      }
      try {
        const list = await this.api.ayahTiming(surahId, readId, slug, reciterId, moshafId);
        await this.cache.write(CACHE.TIMING, key, JSON.stringify(list));
        return ayahTimingToDomain(list, readId, surahId);
      } catch {
        return null;
      }
    });
  }

  /** Warm the cache for the next surah while the current one plays. */
  prefetch(
    readId: number,
    surahId: number,
    slug?: string,
    reciterId?: number,
    moshafId?: number,
  ): void {
    void this.timingFor(readId, surahId, slug, reciterId, moshafId);
  }
}
